pkg/proxy/userspace/proxysocket.go

```diff
@@ -216,61 +239,57 @@ func (udp *udpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
 			glog.Errorf("ReadFrom failed, exiting ProxyLoop: %v", err)
 			break
 		}
-		// If this is a client we know already, reuse the connection and goroutine.
-		svrConn, err := udp.getBackendConn(myInfo.activeClients, cliAddr, proxier, service, myInfo.timeout)
-		if err != nil {
-			continue
-		}
-		// TODO: It would be nice to let the goroutine handle this write, but we don't
-		// really want to copy the buffer.  We could do a pool of buffers or something.
-		_, err = svrConn.Write(buffer[0:n])
-		if err != nil {
-			if !logTimeout(err) {
-				glog.Errorf("Write failed: %v", err)
-				// TODO: Maybe tear down the goroutine for this client/server pair?
+
+		func(udp *udpProxySocket, activeClients *clientCache, buffer *[bufferSize]byte, n int, cliAddr net.Addr, proxier *Proxier, service proxy.ServicePortName, timeout time.Duration) {
+			activeClients.mu.Lock()
+			defer activeClients.mu.Unlock()
+
+			// If this is a client we know already, reuse the connection and goroutine.
+			svrConn, found := activeClients.clients[cliAddr.String()]
+			if !found {
+				// TODO: This could spin up a new goroutine to make the outbound connection,
+				// and keep accepting inbound traffic.
+				glog.V(3).Infof("New UDP connection from %s", cliAddr)
+				go connectAndProxy(service, &fakeUDPConn{udp}, cliAddr, "udp", proxier, func(in, svrConn net.Conn) {
+					if err := svrConn.SetDeadline(time.Now().Add(timeout)); err != nil {
+						glog.Errorf("SetDeadline failed: %v", err)
+						return
+					}
+					activeClients.clients[cliAddr.String()] = svrConn
+					writeData(svrConn, buffer, n, timeout)
```

Comment:
I think this is a problem.  buffer is in the ProxyLoop scope, and will be reused.  Since this runs in a goroutine now, it will race - the ProxyLoop might receive another packet before the old packet is written to the backend.  I think you have to block on a mutex or make a copy of the buffer.